### PR TITLE
ADT: Add basic component metadata type

### PR DESCRIPTION
**Pull request by drwill-ms**
Component sample output:
```
===COMPONENT SAMPLE===

Created models dtmi:com:samples:ComponentModel;1810 and dtmi:com:samples:TempModel;9396.
Created digital twin 'sampleTwin1835'.
Retrieved and deserialized digital twin sampleTwin1835:
        ETag: W/"fde79216-1481-4f8e-a6b7-c4f577ff1561"
        Prop1: Value1
        Prop2: 987
        Component1 metadata: {"ComponentProp1":{"lastUpdateTime":"2020-10-27T20:28:34.0563558Z"},"ComponentProp2":{"lastUpdateTime":"2020-10-27T20:28:34.0563558Z"}}
        Component1.Prop1: Component value 1
        ComponentProp2: 123
Created digital twin 'sampleTwin4104'.
Retrieved and deserialized digital twin sampleTwin4104:
        ETag: W/"43b005f6-43ff-4942-b584-efada0c2d933"
        Prop1: Prop1 val
        Prop2: 987
        ComponentProp1: Component prop1 val last updated 10/27/2020 8:28:34 PM +00:00
        ComponentProp2: 123
Updated component for digital twin 'sampleTwin1835'.
Retrieved component for digital twin 'sampleTwin1835'.
```

### CODE REVIEW

**Review by timtay-microsoft — APPROVED** [review]

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinComponentMetadata.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    /// For more samples, see <see href="https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/digitaltwins/Azure.DigitalTwins.Core/samples">our repo samples</see>.
+    /// </remarks>
+    public class BasicDigitalTwinComponentMetadata
+    {
+        /// <summary>
+        /// This field will contain metadata about changes on properties on a component.
+        /// For your convenience, the value can be deserialized into <see cref="BasicDigitalTwinPropertyMetadata"/>.
+        /// </summary>
+        [JsonExtensionData]
+        public IDictionary<string, object> PropertyMetadata { get; set; } = new Dictionary<string, object>();
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinComponentMetadata.cs:23]: We didn't come up with a name for this dictionary. Thoughts @johngallardo?

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinComponentMetadata.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Text.Json.Serialization;
+
+namespace Azure.DigitalTwins.Core
+{
+    /// <summary>
+    /// An optional, helper class for deserializing a digital twin.
```
Comment (barustum) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinComponentMetadata.cs:10]: Maybe reword to 'deserializing a digital twin component metadata object'

**Review by barustum — APPROVED** [review]

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinComponentMetadata.cs (hunk 1 above)
Comment (johngallardo) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinComponentMetadata.cs:23]: I think PropertyMetadata is fine.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinComponent.cs [hunk 3]
```diff
@@ -22,7 +22,7 @@ public class BasicDigitalTwinComponent
         /// The metadata property, required on a component to identify as one.
         /// </summary>
         [JsonPropertyName(DigitalTwinsJsonPropertyNames.DigitalTwinMetadata)]
-        public object Metadata { get; set; } = new object();
+        public BasicDigitalTwinComponentMetadata Metadata { get; set; } = new BasicDigitalTwinComponentMetadata();
```
Comment (johngallardo) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinComponent.cs:25]: Rather than having a type here. I think we can do something like:

```
[JsonPropertyName(DigitalTwinsJsonPropertyNames.DigitalTwinMetadata)]
IDictionary<string, DigitalTwinPropertyMetadata> Metadata { get; set; } = new Dictinoary<string, DigitalTwinPropertyMetadata>();
```

This will allow us to strongly type the property metadata that exists on the components.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinComponentMetadata.cs (hunk 1 above)
Comment (johngallardo) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinComponentMetadata.cs:23]: But... see my comment in BasicDigitalTwinComponent --- I think we can actually get rid of this type.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinMetadata.cs [hunk 4]
```diff
@@ -19,5 +20,12 @@ public class BasicDigitalTwinMetadata
         /// </summary>
         [JsonPropertyName(DigitalTwinsJsonPropertyNames.MetadataModel)]
         public string ModelId { get; set; }
+
+        /// <summary>
+        /// This field will contain metadata about changes on properties on the digital twin.
+        /// For your convenience, the value can be deserialized into <see cref="BasicDigitalTwinPropertyMetadata"/>.
+        /// </summary>
+        [JsonExtensionData]
```
Comment (johngallardo) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinMetadata.cs:28]: AFAIK - the limitation driving IDictionary<string, object> is a limitation in JsonExtensionData. Did we consider providing a custom attribute on BasicDigitalTwinMetadata which would know how to deserialize these elements? This would allow us to expose

IDictionary<string, DigitalTwinPropertyMetadata>.

And eliminate a bunch of serialization converting that customers would need to do.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinPropertyMetadata.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    /// <summary>
+    /// Contain metadata about changes on properties on a digital twin or component.
+    /// </summary>
+    public class BasicDigitalTwinPropertyMetadata
+    {
+        /// <summary>
+        /// The date and time the property was last updated.
+        /// </summary>
+        [JsonPropertyName(DigitalTwinsJsonPropertyNames.MetadataPropertyLastUpdateTime)]
+        public DateTimeOffset LastUpdatedOn { get; set; }
```
Comment (johngallardo) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinPropertyMetadata.cs:18]: Why not LastUpdateTime? Did we pick a bad property name in the service?

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinMetadata.cs (hunk 4 above)
Comment (johngallardo) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinMetadata.cs:28]: Something like:
```
using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ConsoleApp6
{
    [JsonConverter(typeof(DigitalTwinMetadataJsonConverter))]
    class DigitalTwinMetadata
    {
        public string ModelId { get; set; }

        public IDictionary<string, DigitalTwinPropertyMetadata> PropertyMetadata = new Dictionary<string, DigitalTwinPropertyMetadata>();
    }

    class DigitalTwinPropertyMetadata
    {
        [JsonPropertyName("lastUpdateTime")]
        public DateTimeOffset? LastUpdateTime { get; set; }
    }

    class DigitalTwinMetadataJsonConverter : JsonConverter<DigitalTwinMetadata>
    {
        public override DigitalTwinMetadata Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            if (reader.TokenType == JsonTokenType.Null)
            {
                return null;
            }
            else if (reader.TokenType == JsonTokenType.StartObject)
            {
                var metadata = new DigitalTwinMetadata();
                reader.Read(); // Advance into our object.
                while (reader.TokenType != JsonTokenType.EndObject) // Until we reach the end of the object we began reading
                {
                    if (reader.TokenType == JsonTokenType.PropertyName)
                    {
                        if (reader.GetString() == "$model")
                        {
                            reader.Read(); // advance to the string
                            metadata.ModelId = JsonSerializer.Deserialize<string>(ref reader, options);
                            reader.Read(); // Consume the string we deserialized
                        }
                        else
                        {
                            var propertyName = reader.GetString();
                            reader.Read(); // Should place us on StartObject
                            if (reader.TokenType == JsonTokenType.StartObject)
                            {
                                metadata.PropertyMetadata[propertyName] = JsonSerializer.Deserialize<DigitalTwinPropertyMetadata>(ref reader, options);
                                reader.Read(); // Consume EndObject
                            }
                        }
                    }
                }
                return metadata;
            }
            else
            {
                throw new Exception($"Unknown token type {reader.TokenType} at index {reader.TokenStartIndex}.");
            }
        }

        public override void Write(Utf8JsonWriter writer, DigitalTwinMetadata value, JsonSerializerOptions options)
        {
            writer.WriteStartObject();
            writer.WriteString("$model", value.ModelId);
            foreach(var p in value.PropertyMetadata)
            {
                writer.WritePropertyName(p.Key);
                JsonSerializer.Serialize<DigitalTwinPropertyMetadata>(writer, p.Value, options);
            }
            writer.WriteEndObject();
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            var metadata = new DigitalTwinMetadata
            {
                ModelId = "mymodel",
                PropertyMetadata = new Dictionary<string, DigitalTwinPropertyMetadata>()
                {
                    { "p1", new DigitalTwinPropertyMetadata { LastUpdateTime = DateTimeOffset.Now } },
                    { "p2", new DigitalTwinPropertyMetadata { LastUpdateTime = DateTimeOffset.Now.Subtract(TimeSpan.FromDays(1)) } },
                }
            };

            var ser = JsonSerializer.Serialize(metadata);
            Console.WriteLine(ser);

            // round trip
            var roundTripped = JsonSerializer.Deserialize<DigitalTwinMetadata>(ser);
            Console.WriteLine($"ModelId - {roundTripped.ModelId}");
            foreach(var v in roundTripped.PropertyMetadata)
            {
                Console.WriteLine($"{v.Key} - LastUpdated @ {v.Value.LastUpdateTime}");
            }
        }
    }
}
```

File: sdk/digitaltwins/Azure.DigitalTwins.Core/samples/DigitalTwinsClientSample/CustomDigitalTwin.cs [hunk 6]
```diff
@@ -36,7 +37,7 @@ internal class MyCustomComponent
         /// A component must have a property named $metadata with no properties to be distinguished from other properties as a component.
```
[the remark this answers is not in the document]
Comment (drwill-ms) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/samples/DigitalTwinsClientSample/CustomDigitalTwin.cs:37]: true, it is out of date.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/samples/DigitalTwinsClientSample/CustomDigitalTwin.cs [hunk 7]
```diff
@@ -36,7 +37,7 @@ internal class MyCustomComponent
...
         /// </summary>
         [JsonPropertyName(DigitalTwinsJsonPropertyNames.DigitalTwinMetadata)]
-        public object Metadata { get; set; } = new object();
+        public Dictionary<string, BasicDigitalTwinPropertyMetadata> Metadata { get; set; } = new Dictionary<string, BasicDigitalTwinPropertyMetadata>();
```
[the remark this answers is not in the document]
Comment (drwill-ms) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/samples/DigitalTwinsClientSample/CustomDigitalTwin.cs:40]: ok

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinComponent.cs (hunk 3 above)
Comment (drwill-ms) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinComponent.cs:25]: I think that works. I'll give it a shot.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinPropertyMetadata.cs (hunk 5 above)
Comment (timtay-microsoft) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinPropertyMetadata.cs:18]: Azure SDK board prefers we use this style of naming for timestamp fields. For instance, we use "UploadedOn" rather than "UploadTime" in the digital twins model data class

Comment (johngallardo) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinPropertyMetadata.cs:18]: Even if that doesn't align with the service API? Is this language specific?

Comment (johngallardo) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinPropertyMetadata.cs:18]: Also - is there a recommendation on the MetadataPropertyLastUpdateTime naming? Should the name of that constant align with the name used by the service, or the name used in the projected API?

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinComponent.cs (hunk 3 above)
Comment (johngallardo) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinComponent.cs:25]: Alternatively we can use a custom converter to avoid the limitations of JsonExtensionData.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinPropertyMetadata.cs (hunk 5 above)
Comment (drwill-ms) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinPropertyMetadata.cs:18]: Yes, the Azure SDK board says they prefer the suffix On for date/time properties.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinMetadata.cs (hunk 4 above)
Comment (drwill-ms) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinMetadata.cs:28]: We can't control which json serialization settings the user uses.

Comment (johngallardo) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwinMetadata.cs:28]: How is attributing with a JsonConverter any different than attributing with JsonExtensionData? In either case, the owner of the JsonSerializerOptions can configure the set of JsonConverter objects that are used for any type... this is just controlling the defaults, and providing the best possible default seems like it would have value.
